### DISCUSSION TITLE
integration question

### DISCUSSION THREAD

**Discussion by foxnight2** [nlohmann/json#4251] [Q&A]
https://github.com/nlohmann/json?tab=readme-ov-file#integration

I saw the embed method, it is great.

If I doesnt want to place the entire source tree in a subdirectory and call add_subdirectory() in your CMakeLists.txt file.
What should I do ?

Can I replace the method with this

First I put the https://github.com/nlohmann/json/releases into my project.

where ${NLOHMANNJSON_INCLUDE_DIRS} path is at my project
${NLOHMANNJSON_INCLUDE_DIRS} = deps/nlohmann_json/include

```
# CMakeLists.txt

find_path(
	NLOHMANNJSON_INCLUDE_DIRS
	NAMES nlohmann/json.hpp
	PATHS "deps/nlohmann_json/include"
)

if(json_FOUND)
  if(NOT TARGET nlohmann_json::nlohmann_json)
    add_library(nlohmann_json::nlohmann_json INTERFACE IMPORTED)
    set_target_properties(nlohmann_json::nlohmann_json PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${NLOHMANNJSON_INCLUDE_DIRS}")
  endif()
endif()

add_executable(test_app ${CMAKE_CURRENT_SOURCE_DIR}/main.cpp)
target_link_libraries(test_app PRIVATE nlohmann_json::nlohmann_json)
```

here is my project structure
`deps/nlohmann_json/include/nlohmann`
```
- project
  - .gitignore
  - README.md
  - LICENSE.md
  - CMakeLists.txt
  - cmake
    - FindSomeLib.cmake
    - something_else.cmake
  - include
    - project
      - lib.hpp
  - src
    - CMakeLists.txt
    - lib.cpp
  - apps
    - CMakeLists.txt
    - app.cpp
  - tests
    - CMakeLists.txt
    - testlib.cpp
  - docs
    - CMakeLists.txt
  - extern
    - googletest
  - scripts
    - helper.py
  - deps
     - nlohmann_json
        - include
            - nlohmann
```

**Comment by nlohmann** [comment]
Since nlohmann/json is header-only, you don't need the CMakeLists.txt-based add_subdirectory() target at all — your manual approach (find_path + an INTERFACE IMPORTED target with INTERFACE_INCLUDE_DIRECTORIES) is valid and is essentially what add_subdirectory() would set up for you anyway. An even simpler option for your layout is to skip the imported target and just call target_include_directories(test_app PRIVATE deps/nlohmann_json/include) directly. See the README's 'Embedded' and 'CMake' integration sections for the officially supported add_subdirectory()/find_package()/FetchContent() variants if you'd rather not hand-roll the imported target.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
